Import the Kaggle Dataset with the artist information

We will be focusing on Genre, Style, and Era (Date)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
%cd drive/MyDrive/CS 348K Final Project

/content/drive/MyDrive/CS 348K Final Project


In [ ]:
!ls

'CS 348K Final Project Notebook.ipynb'	'Kaggle JSON Folder'   train_3.zip
 date_classification.csv		 train_1	       train_info.csv
 date_classification.h5			 train_1.zip	       train_info.csv.zip
 genre_classification_4_categories.h5	 train_2	      'Training Images'
 genre_classification.csv		 train_2.zip	      'Training Images - 256'
 genre_classification.h5		 train_3	      'Training Images New'


In [ ]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = 'Kaggle JSON Folder'

In [ ]:
!pip install -q kaggle

In [ ]:
!kaggle competitions download -c painter-by-numbers

404 - Not Found - no gcs url


Unzipping files

In [ ]:
!ls

'CS 348K Final Project Notebook.ipynb'	'Kaggle JSON Folder'   train_3.zip
 date_classification.csv		 train_1	       train_info.csv
 date_classification.h5			 train_1.zip	       train_info.csv.zip
 genre_classification_4_categories.h5	 train_2	      'Training Images'
 genre_classification.csv		 train_2.zip	      'Training Images - 256'
 genre_classification.h5		 train_3	      'Training Images New'


In [ ]:
import zipfile

# Path to the zip file
zip_path_1 = 'train_1.zip'
zip_path_2 = 'train_info.csv.zip'
zip_path_3 = 'train_2.zip'
zip_path_4 = 'train_3.zip'

# Open the zip file in read mode


with zipfile.ZipFile(zip_path_1, 'r') as zip_ref_1:
    # Extract all the contents into the directory specified
    zip_ref_1.extractall()


print("Files extracted successfully!")

Files extracted successfully!


In [ ]:
!ls

'CS 348K Final Project Notebook.ipynb'	'Kaggle JSON Folder'   train_3.zip
 date_classification.csv		 train_1	       train_info.csv
 date_classification.h5			 train_1.zip	       train_info.csv.zip
 genre_classification_4_categories.h5	 train_2	      'Training Images'
 genre_classification.csv		 train_2.zip	      'Training Images - 256'
 genre_classification.h5		 train_3	      'Training Images New'


In [ ]:
import os

In [ ]:
print(len(os.listdir("train_1")))



11025


In [ ]:
print(len(os.listdir("Training Images - 256")))

19240


In [ ]:
print(len(os.listdir("train_2")))

8476


In [ ]:
print(len(os.listdir("train_3")))

8531


Total should be 19242 (or somewhere around there if there are errors).

In [ ]:
import zipfile
zip_path_2 = "train_info.csv.zip"
with zipfile.ZipFile(zip_path_2, 'r') as zip_ref_2:
    # Extract all the contents into the directory specified
    zip_ref_2.extractall()

In [ ]:
# DONT NEED TO RUN THIS ANYMORE, RUN THE CELL BELOW IT INSTEAD.


my_images_1 = os.listdir("train_1")  # Your list of filenames goes here
my_images_2 = os.listdir("train_2")
my_images_3 = os.listdir("train_3")
# Convert list to a set for faster lookup
my_images_1.extend(my_images_2)
my_images_1.extend(my_images_3)

my_images_set = set(my_images_1)

# Remove image that has a problem: 101947.jpg
my_images_set.remove("101947.jpg")

In [ ]:
import os

my_images_set = os.listdir("Training Images - 256")

In [ ]:
print(len(my_images_set))

19240


In [ ]:
import csv
import os

# Assume my_images is a list of filenames you are interested in

# Path to the CSV file
csv_file_path = 'train_info.csv'

# Dictionary where the filename is the key, only if the filename is in my_images
filtered_data_by_filename = {}

with open(csv_file_path, mode='r', newline='', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    for row in reader:
        filename = row['filename']
        if filename in my_images_set:
            # Create a dictionary for each relevant filename
            filtered_data_by_filename[filename] = {
                'artist': row['artist'],
                'title': row['title'],
                'style': row['style'],
                'genre': row['genre'],
                'date': row['date']
            }

# print("Filtered dictionary by filename:", filtered_data_by_filename)


In [ ]:
print(len(filtered_data_by_filename))

19240


In [ ]:
# Dictionary with the artist as the key, and list of tuples as values
data_by_artist = {}

for filename, details in filtered_data_by_filename.items():
    artist = details['artist']
    # Create a tuple containing the filename, title, style, genre, and date
    file_info = (filename, details['title'], details['style'], details['genre'], details['date'])

    # Check if the artist already exists in the dictionary
    if artist in data_by_artist:
        data_by_artist[artist].append(file_info)
    else:
        data_by_artist[artist] = [file_info]

print("Dictionary by artist:", data_by_artist)

Dictionary by artist: {'5b39c876740bfc1cfaf544721c43cac3': [('102257.jpg', 'Uriel', 'Color Field Painting', 'abstract', '1955.0'), ('100820.jpg', 'Eve ', 'Color Field Painting', 'abstract', '')], '46200eee40e22039aef7a5944bb445c9': [('100478.jpg', 'Cover design of the first publication of the Association of Austrian Artists Secession', 'Art Nouveau (Modern)', 'design', '1897'), ('102149.jpg', 'Draft drawings for the breakfast room of the apartment Eisler Terramare, ladies desk chair with retractable', 'Art Nouveau (Modern)', 'design', '1903'), ('101111.jpg', 'Holy Spirit Church in Düsseldorf, design for the right side wall', 'Art Nouveau (Modern)', 'design', '1907'), ('101602.jpg', 'November', 'Art Nouveau (Modern)', 'genre painting', '1902'), ('100482.jpg', "Sketch of emblem to 'Ver Sacrum'", 'Art Nouveau (Modern)', 'design', '1898'), ('100808.jpg', 'A decorative stain in red and green', 'Art Nouveau (Modern)', 'design', '1897')], '74a1dad53214d04514269a188f16683a': [('100829.jpg', 'T

In [ ]:
print(len(data_by_artist))

NameError: name 'data_by_artist' is not defined

In [ ]:
# Dictionary with the style as the key, and list of tuples as values
data_by_style = {}

for filename, details in filtered_data_by_filename.items():
    style = details['style']
    # Create a tuple containing the filename, artist, title, genre, and date
    file_info = (filename, details['artist'], details['title'], details['genre'], details['date'])

    # Check if the style already exists in the dictionary
    if style in data_by_style:
        data_by_style[style].append(file_info)
    else:
        data_by_style[style] = [file_info]

print("Dictionary by style:", data_by_style)

Dictionary by style: {'Color Field Painting': [('102257.jpg', '5b39c876740bfc1cfaf544721c43cac3', 'Uriel', 'abstract', '1955.0'), ('100820.jpg', '5b39c876740bfc1cfaf544721c43cac3', 'Eve ', 'abstract', ''), ('3108.jpg', 'cea636ca5aacb6df67baeead1500eeb5', 'Ocean Park No. 43', 'abstract', ''), ('39891.jpg', 'bc0ddf03c667c5edb17982be481ef360', 'The Snail', 'abstract', '1953'), ('30738.jpg', 'cea636ca5aacb6df67baeead1500eeb5', 'Ocean Park No. 40', 'abstract', ''), ('27025.jpg', '7b016984b86b58f83977299591cf4e38', 'Big Red', 'abstract', '1953'), ('38603.jpg', 'd1ec00902766b952ec62d98953e135ec', 'No. 16 (Red, White, and Brown)', 'abstract', '1957'), ('36591.jpg', '5a98f9e90e9ee706df88fcedab15677e', 'Cross Over ', 'abstract', '1974.0'), ('23655.jpg', '4f616f2a876dad7850ce66f465503a5d', 'Untitled from the Shapes of Colour', 'abstract', '1978'), ('22818.jpg', 'db85b84c1c01fa143b9294f00f7acbeb', 'Two Part Blue', 'abstract', '1964'), ('100876.jpg', 'db85b84c1c01fa143b9294f00f7acbeb', 'Black Popco

In [ ]:
print(len(data_by_style))

average_images_per_style = 0
for style_type in data_by_style:
  average_images_per_style += len(data_by_style[style_type])

average_images_per_style = average_images_per_style / len(data_by_style)
print("The average number of images per style is: " + str(average_images_per_style))

132
The average number of images per style is: 145.75757575757575


In [ ]:
# Dictionary with the genre as the key, and list of tuples as values
data_by_genre = {}

for filename, details in filtered_data_by_filename.items():
    genre = details['genre']
    # Create a tuple containing the filename, artist, title, style, and date
    file_info = (filename, details['artist'], details['title'], details['style'], details['date'])

    # Check if the genre already exists in the dictionary
    if genre in data_by_genre:
        data_by_genre[genre].append(file_info)
    else:
        data_by_genre[genre] = [file_info]

# print("Dictionary by genre:", data_by_genre)

In [ ]:
print(len(data_by_genre))

42


In [ ]:
top_keys = sorted(data_by_genre, key=lambda k: len(data_by_genre[k]), reverse=True)[:4]

data_by_genre_top_4 = {key: data_by_genre[key] for key in top_keys}


In [ ]:
print(len(data_by_genre_top_4))

4


In [ ]:
top_n_keys = ['portrait', 'landscape', 'genre painting', 'abstract']

total_images = 0
for key in top_n_keys:
  total_images += len(data_by_genre[key])

print(total_images)

10313


In [ ]:
print(len(data_by_genre))

average_per_class = 0

for keys in data_by_genre.keys():
  average_per_class += len(data_by_genre[keys])
  print(len(data_by_genre[keys]))

average_per_class = average_per_class / len(data_by_genre.keys())

print("The average per class is: " + str(average_per_class))

42
1722
365
30
2616
1391
606
251
97
435
188
586
2787
276
384
3188
481
667
964
51
119
45
37
163
314
56
48
68
318
321
421
31
17
2
43
20
40
8
6
32
29
14
3
The average per class is: 458.0952380952381


There are on average 458 images per each class in the genre class when we incorporate train_1, train_2, and train_3.

In [ ]:
from collections import defaultdict
import math

# Initialize dictionary to hold categorized data
data_by_date_range = defaultdict(list)

# Collect all dates from the filtered data to determine the min and max
dates = [int(details['date']) for filename, details in filtered_data_by_filename.items() if details['date'].isdigit()]

if not dates:
    print("No valid date data found in the filtered dictionary.")
    exit()

# Determine adjusted date range start and end
min_year = min(dates)
max_year = max(dates)
base_year = min_year - (min_year % 50)
end_year = max_year + (50 - max_year % 50)

# Function to assign date ranges
def determine_date_range(year):
    if year < base_year:
        return f"Before {base_year}"
    elif year > end_year:
        return f"After {end_year}"
    range_start = ((year - base_year) // 50) * 50 + base_year
    range_end = range_start + 49
    return f"{range_start}-{range_end}"

# Categorize entries by date range
for filename, details in filtered_data_by_filename.items():
    try:
        year = int(details['date'])
        date_range = determine_date_range(year)
        file_info = (filename, details['artist'], details['title'], details['style'], details['genre'], details['date'])
        data_by_date_range[date_range].append(file_info)
    except ValueError:
        continue  # Skip entries with invalid dates

# Output the data categorized by date range
'''
for range_key, files in data_by_date_range.items():
    print(f"Date Range {range_key}:")
    for file in files:
        print(file)
'''

'\nfor range_key, files in data_by_date_range.items():\n    print(f"Date Range {range_key}:")\n    for file in files:\n        print(file)\n'

This code splits the data into exactly 5 date ranges

In [ ]:
from collections import defaultdict
import math

# Initialize dictionary to hold categorized data
data_by_date_range = defaultdict(list)

# Collect all dates from the filtered data to determine the min and max
dates = [int(details['date']) for filename, details in filtered_data_by_filename.items() if details['date'].isdigit()]

if not dates:
    print("No valid date data found in the filtered dictionary.")
    exit()

# Determine the range of dates and create 5 equal intervals
min_year = min(dates)
max_year = max(dates)
interval_length = math.ceil((max_year - min_year) / 5)

# Function to assign date ranges
def determine_date_range(year):
    if year < min_year:
        return f"Before {min_year}"
    elif year > max_year:
        return f"After {max_year}"
    range_index = (year - min_year) // interval_length
    range_start = min_year + range_index * interval_length
    range_end = range_start + interval_length - 1
    # Adjust the last interval to not exceed the max_year
    if range_end > max_year:
        range_end = max_year
    return f"{range_start}-{range_end}"

# Categorize entries by date range
for filename, details in filtered_data_by_filename.items():
    try:
        year = int(details['date'])
        date_range = determine_date_range(year)
        file_info = (filename, details['artist'], details['title'], details['style'], details['genre'], details['date'])
        data_by_date_range[date_range].append(file_info)
    except ValueError:
        continue  # Skip entries with invalid dates

# Output the data categorized by date range
for range_key, files in data_by_date_range.items():
    print(f"{range_key}: {len(files)} files")


1845-2010: 6328 files
1513-1678: 631 files
1679-1844: 783 files
1347-1512: 324 files
1181-1346: 46 files


In [ ]:
import random

def sample_images(image_list, num_samples=700):
    return random.sample(image_list, min(num_samples, len(image_list)))


In [ ]:
new_data_by_date_range = {
    '1845-2010': sample_images(data_by_date_range['1845-2010'], 700),
    '1513-1678': data_by_date_range['1513-1678'],
    '1679-1844': data_by_date_range['1679-1844'],
    '1347-1512': data_by_date_range['1347-1512'],
    '1181-1346': data_by_date_range['1181-1346']
}

In [ ]:
# Output the data categorized by date range
for range_key, files in new_data_by_date_range.items():
    print(f"{range_key}: {len(files)} files")


1845-2010: 700 files
1513-1678: 631 files
1679-1844: 783 files
1347-1512: 324 files
1181-1346: 46 files


In [ ]:
print(len(data_by_date_range['1850-1899']))

2268


In [ ]:
print(data_by_date_range.keys())
print(len(data_by_date_range))

dict_keys(['1850-1899', '1900-1949', '1950-1999', '1500-1549', '1800-1849', '1600-1649', '1750-1799', '1450-1499', '1550-1599', '1400-1449', '1700-1749', '1650-1699', '1350-1399', '1250-1299', '1300-1349', '2000-2049', '1150-1199'])
17


In [ ]:
import os
from PIL import Image

# Define the path to your image folder
source_folder = 'train_1'

# Dictionary containing filenames as keys (assuming it's already loaded)
# filtered_data_by_filename

# Initialize sums and counts
total_width = 0
total_height = 0
count = 0

# Loop through all files in the source folder

for filename in os.listdir(source_folder):
    if count == 50:
      break
    if filename in filtered_data_by_filename:
        file_path = os.path.join(source_folder, filename)
        try:
            # Open the image and get its dimensions
            with Image.open(file_path) as img:
                width, height = img.size
                total_width += width
                total_height += height
                count += 1
                print(f"Processed {filename}: Width = {width}, Height = {height}")
        except IOError as e:
            print(f"Error processing {filename}: {e}")

# Calculate the average dimensions
if count > 0:
    average_width = total_width / count
    average_height = total_height / count
    print(f"Average Width: {average_width}")
    print(f"Average Height: {average_height}")
else:
    print("No images were processed.")


Processed 101441.jpg: Width = 977, Height = 1126
Processed 101443.jpg: Width = 567, Height = 685
Processed 101439.jpg: Width = 1184, Height = 990
Processed 101440.jpg: Width = 678, Height = 1000
Processed 101605.jpg: Width = 805, Height = 1024
Processed 101609.jpg: Width = 595, Height = 1000
Processed 101604.jpg: Width = 563, Height = 517
Processed 101608.jpg: Width = 1800, Height = 1324
Processed 10161.jpg: Width = 563, Height = 375
Processed 101571.jpg: Width = 850, Height = 1157
Processed 101612.jpg: Width = 847, Height = 944
Processed 101538.jpg: Width = 2626, Height = 3472
Processed 101511.jpg: Width = 374, Height = 450
Processed 101686.jpg: Width = 3344, Height = 2796
Processed 101483.jpg: Width = 1682, Height = 1799
Processed 101585.jpg: Width = 708, Height = 1000
Processed 101455.jpg: Width = 3504, Height = 2383
Processed 101521.jpg: Width = 1087, Height = 1114
Processed 101885.jpg: Width = 960, Height = 1379
Processed 101886.jpg: Width = 583, Height = 900
Processed 101883.jpg:

In [ ]:
standard_image_width = 256
standard_image_height = 256

In [ ]:
import os
from PIL import Image

# Define paths
source_folder = 'train_1'
destination_folder = 'Training Images'
os.makedirs(destination_folder, exist_ok=True)  # Create the destination folder if it doesn't exist

# Set the desired size
desired_size = (standard_image_width, standard_image_height)

# Dictionary containing filenames as keys (you need to have this preloaded)
# filtered_data_by_filename

# Loop through all files in the source folder
count = 0
for filename in error_filenames:
# for filename in os.listdir(source_folder):
    if filename in filtered_data_by_filename:
        # Construct the full file path
        file_path = os.path.join(source_folder, filename)
        try:
            # Open the image
            with Image.open(file_path) as img:
                # Convert image to RGB if it's not
                if img.mode != 'RGB':
                    img = img.convert('RGB')
                # Resize the image
                img_resized = img.resize(desired_size, Image.ANTIALIAS)
                # Save the resized image to the destination folder
                img_resized.save(os.path.join(destination_folder, filename), 'JPEG')  # Specify format explicitly
                print(f"Resized and saved {filename}")
        except IOError as e:
            print(f"Error processing {filename}: {e}")

        count += 1
        if count % 10 == 0:
            print("Completed resizing " + str(count) + "/2236 images")  # Updated with your specific total if known

print("Image processing complete.")


Error processing 101947.jpg: image file is truncated (80 bytes not processed)
Resized and saved 102290.jpg
Resized and saved 100498.jpg
Resized and saved 101332.jpg
Image processing complete.


<ipython-input-46-6bfe6711c97d>:29: DeprecationWarning: ANTIALIAS is deprecated and will be removed in Pillow 10 (2023-07-01). Use LANCZOS or Resampling.LANCZOS instead.
  img_resized = img.resize(desired_size, Image.ANTIALIAS)


Adding two more training sets of images to the directory:

Loading all the images to "Training Images"

In [ ]:
print(len(os.listdir("Training Images")))

19240


In [ ]:
print(len(os.listdir("train_1")) + len(os.listdir("train_2")) + len(os.listdir("train_3")))

28032


In [ ]:
import os
from PIL import Image

# Define source and destination directories
source_dirs = ['train_1', 'train_2', 'train_3']
destination_dir = 'Training Images'

# Ensure the destination directory exists
if not os.path.exists(destination_dir):
    os.makedirs(destination_dir)

# Define the new for the images
new_size = (64,64)  # Example size, adjust as needed

error_files = []  # To store files which caused an error when trying to resize

images_completed = []

count = 0
for source_dir in source_dirs:
    for filename in os.listdir(source_dir):
        if filename.endswith(('jpg', 'jpeg', 'png', 'bmp', 'gif')):  # Include other formats if needed
            source_path = os.path.join(source_dir, filename)
            destination_path = os.path.join(destination_dir, filename)

            try:
              # Open the image
              with Image.open(source_path) as img:
                  # Convert image to RGB if it's in RGBA mode
                  if img.mode in ('RGBA', 'P'):
                      img = img.convert('RGB')

                  # Resize the image
                  img = img.resize(new_size, Image.ANTIALIAS)

                  # Save the resized image to the destination path
                  img.save(destination_path)
            except Exception as e:  # General exception to catch any error
                  # Append the file that caused the error
                  error_files.append(source_path)
                  # Optionally print an error message or log it
                  print(f"Error processing {source_path}: {e}")

        images_completed.append(filename)
        count += 1

        if count % 10 == 0:

          print("Resized " + str(count) + "/28032")

print("Images have been converted (if needed), resized, and moved successfully.")
print("These are the images which caused errors when trying to resize:")
print(error_files)

<ipython-input-26-284a692675a2>:34: DeprecationWarning: ANTIALIAS is deprecated and will be removed in Pillow 10 (2023-07-01). Use LANCZOS or Resampling.LANCZOS instead.
  img = img.resize(new_size, Image.ANTIALIAS)


Resized 10/28032
Resized 20/28032
Resized 30/28032
Resized 40/28032
Resized 50/28032
Resized 60/28032
Resized 70/28032
Resized 80/28032
Resized 90/28032
Resized 100/28032
Resized 110/28032
Resized 120/28032
Resized 130/28032
Resized 140/28032
Resized 150/28032
Resized 160/28032
Resized 170/28032
Resized 180/28032
Resized 190/28032
Resized 200/28032
Resized 210/28032
Resized 220/28032
Resized 230/28032
Resized 240/28032
Resized 250/28032
Resized 260/28032
Resized 270/28032
Resized 280/28032


KeyboardInterrupt: 

Flattened Version

In [ ]:
print(type(new_data_by_date_range['1845-2010']))

<class 'list'>


In [ ]:
new_data_by_date_range['1845-2010'][0]

('37083.jpg',
 'a605a99aaac53ee97fb05dd6bbac6c28',
 'The Arab Tent',
 'Romanticism',
 'animal painting',
 '1866')

In [ ]:
import os
from PIL import Image

# Define source and destination directories
source_dirs = ['train_1', 'train_2', 'train_3']
destination_dir = 'Training Images New'

# Ensure the destination directory exists
if not os.path.exists(destination_dir):
    os.makedirs(destination_dir)

# Define the new size for the images
new_size = (64, 64)  # Example size, adjust as needed

# Sample new_data_by_date_range structure

# Flatten the dictionary into a set of filenames

file_set = []
for date_class in new_data_by_date_range:
  current_list_of_image_features = new_data_by_date_range[date_class]
  for i in range(len(current_list_of_image_features)):
    file_set.append(current_list_of_image_features[i][0])

print(file_set)
print(len(file_set))

file_set = set(file_set)

error_files = []  # To store files which caused an error when trying to resize

images_completed = []

count = 0
for source_dir in source_dirs:
    for filename in os.listdir(source_dir):
        if filename in file_set and filename.endswith(('jpg', 'jpeg', 'png', 'bmp', 'gif')):  # Include other formats if needed
            source_path = os.path.join(source_dir, filename)
            destination_path = os.path.join(destination_dir, filename)

            try:
                # Open the image
                with Image.open(source_path) as img:
                    # Convert image to RGB if it's in RGBA mode
                    if img.mode in ('RGBA', 'P'):
                        img = img.convert('RGB')

                    # Resize the image
                    img = img.resize(new_size, Image.ANTIALIAS)

                    # Save the resized image to the destination path
                    img.save(destination_path)
            except Exception as e:  # General exception to catch any error
                # Append the file that caused the error
                error_files.append(source_path)
                # Optionally print an error message or log it
                print(f"Error processing {source_path}: {e}")

            images_completed.append(filename)
            count += 1

            if count % 10 == 0:
                print(f"Resized {count}/2484")

print("Images have been converted (if needed), resized, and moved successfully.")
print("These are the images which caused errors when trying to resize:")
print(error_files)


['37083.jpg', '37629.jpg', '24541.jpg', '23178.jpg', '31251.jpg', '3578.jpg', '31113.jpg', '102053.jpg', '38439.jpg', '102090.jpg', '33018.jpg', '100710.jpg', '20429.jpg', '37405.jpg', '37641.jpg', '37510.jpg', '3338.jpg', '36385.jpg', '101769.jpg', '234.jpg', '34138.jpg', '34348.jpg', '33262.jpg', '36628.jpg', '20608.jpg', '31603.jpg', '24502.jpg', '37460.jpg', '3914.jpg', '3999.jpg', '22923.jpg', '25687.jpg', '29366.jpg', '27853.jpg', '25617.jpg', '31304.jpg', '29755.jpg', '36520.jpg', '24034.jpg', '36494.jpg', '23099.jpg', '2936.jpg', '21221.jpg', '100857.jpg', '24603.jpg', '101360.jpg', '21018.jpg', '22730.jpg', '3420.jpg', '35636.jpg', '2956.jpg', '2492.jpg', '35649.jpg', '100256.jpg', '3701.jpg', '32058.jpg', '101545.jpg', '100391.jpg', '23235.jpg', '36205.jpg', '39496.jpg', '22100.jpg', '22028.jpg', '31896.jpg', '21314.jpg', '10243.jpg', '32867.jpg', '25309.jpg', '30299.jpg', '35146.jpg', '20174.jpg', '23119.jpg', '26124.jpg', '3950.jpg', '39719.jpg', '24835.jpg', '24101.jpg', '

<ipython-input-33-c12dba4d9a96>:49: DeprecationWarning: ANTIALIAS is deprecated and will be removed in Pillow 10 (2023-07-01). Use LANCZOS or Resampling.LANCZOS instead.
  img = img.resize(new_size, Image.ANTIALIAS)


Resized 10/2484
Resized 20/2484
Resized 30/2484
Resized 40/2484
Resized 50/2484
Resized 60/2484
Resized 70/2484
Resized 80/2484
Resized 90/2484
Resized 100/2484
Resized 110/2484
Resized 120/2484
Resized 130/2484
Resized 140/2484
Resized 150/2484
Resized 160/2484
Resized 170/2484
Resized 180/2484
Resized 190/2484
Resized 200/2484
Resized 210/2484
Resized 220/2484
Resized 230/2484
Resized 240/2484
Resized 250/2484
Resized 260/2484
Resized 270/2484
Resized 280/2484
Resized 290/2484
Resized 300/2484
Resized 310/2484
Resized 320/2484
Resized 330/2484
Resized 340/2484
Resized 350/2484
Resized 360/2484
Resized 370/2484
Resized 380/2484
Resized 390/2484
Resized 400/2484
Resized 410/2484
Resized 420/2484
Resized 430/2484
Resized 440/2484
Resized 450/2484
Resized 460/2484
Resized 470/2484
Resized 480/2484
Resized 490/2484
Resized 500/2484
Resized 510/2484
Resized 520/2484
Resized 530/2484
Resized 540/2484
Resized 550/2484
Resized 560/2484
Resized 570/2484
Resized 580/2484
Resized 590/2484
Resize

In [ ]:
import os
from PIL import Image

# Specify the directory containing the images
directory_path = 'Training Images'

# Create a list to hold the image sizes
image_sizes = []

# Traverse the directory
count = 0
for filename in os.listdir(directory_path):
    # Construct full file path
    file_path = os.path.join(directory_path, filename)
    # Check if the file is an image
    if filename.lower().endswith(('.png', '.jpg', '.jpeg', '.bmp', '.gif', '.tiff')):
        try:
            # Open the image file
            with Image.open(file_path) as img:
                # Get the size of the image
                size = img.size
                # Append the size to the list
                image_sizes.append((filename, size))
                print(f"{filename}: {size}")
        except IOError:
            print(f"Failed to open or process {filename}")

    count += 1
    if count == 10:
      break

# Optionally, print all sizes or do further processing with image_sizes
print(image_sizes)


31523.jpg: (64, 64)
31524.jpg: (64, 64)
31525.jpg: (64, 64)
31526.jpg: (64, 64)
31527.jpg: (64, 64)
31528.jpg: (64, 64)
31529.jpg: (64, 64)
3153.jpg: (64, 64)
31530.jpg: (64, 64)
31531.jpg: (64, 64)
[('31523.jpg', (64, 64)), ('31524.jpg', (64, 64)), ('31525.jpg', (64, 64)), ('31526.jpg', (64, 64)), ('31527.jpg', (64, 64)), ('31528.jpg', (64, 64)), ('31529.jpg', (64, 64)), ('3153.jpg', (64, 64)), ('31530.jpg', (64, 64)), ('31531.jpg', (64, 64))]


Processing images that had errors

In [ ]:
error_filenames_train_set_1 = ["101947.jpg", "102290.jpg", "100498.jpg", "101332.jpg"]

In [ ]:
error_filenames_train_set_3 = ["train_3/3917.jpg", "train_3/33557.jpg"]

There were no problematic images for train_set_3

In [ ]:
print(len(os.listdir("Training Images")))

19240


In [ ]:
print(len(filtered_data_by_filename))

19240


Remove any images from filtered_data_by_filename that didn't resize and process properly into the "Training Images" directory.

In [ ]:
filtered_data_by_filename.pop("3917.jpg")

{'artist': 'f7b0b8b497d3e489bd836159807ba906',
 'title': 'Cows by a Stream',
 'style': 'Tonalism',
 'genre': 'literary painting',
 'date': ''}

In [ ]:
filtered_data_by_filename.pop("33557.jpg")

{'artist': '3706730fd777d92b658f702c241c45e5',
 'title': 'Cathedral',
 'style': 'Orphism',
 'genre': 'cityscape',
 'date': '1913'}

In [ ]:
print(len(filtered_data_by_filename))

19240


In [ ]:
if len(filtered_data_by_filename) == len(os.listdir("Training Images")):
  print("Success. The database and the Image files align.")

Success. The database and the Image files align.


In [ ]:
import csv

# File paths
input_csv = 'genre_classification.csv'
output_csv = 'genre_classification.csv'  # This can be the same as input_csv to overwrite

# The value to search for in the first column
target_value = '101947.jpg'

# Read the CSV and filter out the row
filtered_rows = []
with open(input_csv, mode='r', newline='') as file:
    reader = csv.reader(file)
    for row in reader:
        if row[0] != target_value:
            filtered_rows.append(row)

# Write the filtered rows back to the CSV
with open(output_csv, mode='w', newline='') as file:
    writer = csv.writer(file)
    writer.writerows(filtered_rows)

print("Row deletion complete.")

Row deletion complete.


In [ ]:
print(len(filtered_data_by_filename))
filtered_data_by_filename.pop("101947.jpg")

2235


KeyError: '101947.jpg'

Preparing CSV for Genre Classification

In [ ]:
!ls

'CS 348K Final Project Notebook.ipynb'	'Kaggle JSON Folder'   train_1.zip      train_info.csv.zip
 genre_classification.csv		 train_1	       train_info.csv  'Training Images'


In [ ]:
print(len(data_by_genre))

NameError: name 'data_by_genre' is not defined

In [ ]:
print(filtered_data_by_filename["102257.jpg"])

{'artist': '5b39c876740bfc1cfaf544721c43cac3', 'title': 'Uriel', 'style': 'Color Field Painting', 'genre': 'abstract', 'date': '1955.0'}


In [ ]:
data_by_genre_top_4_filenames_genre = []

count = 0
for genre in data_by_genre_top_4:
  # print(genre)
  for item in range(len(data_by_genre_top_4[genre])):
    # print(data_by_genre_top_4[genre][item])
    data_by_genre_top_4_filenames_genre.append((data_by_genre_top_4[genre][item][0], genre))


In [ ]:
print(data_by_genre_top_4_filenames_genre)

[('100829.jpg', 'portrait'), ('22404.jpg', 'portrait'), ('32754.jpg', 'portrait'), ('34382.jpg', 'portrait'), ('31655.jpg', 'portrait'), ('25503.jpg', 'portrait'), ('393.jpg', 'portrait'), ('254.jpg', 'portrait'), ('102442.jpg', 'portrait'), ('23382.jpg', 'portrait'), ('35239.jpg', 'portrait'), ('27639.jpg', 'portrait'), ('29007.jpg', 'portrait'), ('3016.jpg', 'portrait'), ('33511.jpg', 'portrait'), ('22248.jpg', 'portrait'), ('39349.jpg', 'portrait'), ('27826.jpg', 'portrait'), ('102518.jpg', 'portrait'), ('2165.jpg', 'portrait'), ('30259.jpg', 'portrait'), ('377.jpg', 'portrait'), ('211.jpg', 'portrait'), ('29558.jpg', 'portrait'), ('31728.jpg', 'portrait'), ('26042.jpg', 'portrait'), ('25842.jpg', 'portrait'), ('20440.jpg', 'portrait'), ('27768.jpg', 'portrait'), ('35275.jpg', 'portrait'), ('26697.jpg', 'portrait'), ('100883.jpg', 'portrait'), ('30276.jpg', 'portrait'), ('21978.jpg', 'portrait'), ('33753.jpg', 'portrait'), ('33311.jpg', 'portrait'), ('35626.jpg', 'portrait'), ('3218

Write csv for genre

In [ ]:
import csv

# Example dictionary
# filtered_data_by_filename

# Path where the CSV will be saved
csv_file_path = 'genre_classification.csv'

# Create and write to the CSV file
with open(csv_file_path, mode='w', newline='') as file:
    writer = csv.writer(file)
    # Write the header
    writer.writerow(['filename', 'genre'])
    # Write data rows
    for filename, genre in data_by_genre_top_4_filenames_genre:
        writer.writerow([filename, genre])

print(f"CSV file has been created at {csv_file_path}.")


CSV file has been created at genre_classification.csv.


Write csv for date ranges

In [ ]:
print(data_by_date_range['1850-1899'])

[('100478.jpg', '46200eee40e22039aef7a5944bb445c9', 'Cover design of the first publication of the Association of Austrian Artists Secession', 'Art Nouveau (Modern)', 'design', '1897'), ('100829.jpg', '74a1dad53214d04514269a188f16683a', 'The pastor', 'Realism', 'portrait', '1879'), ('102518.jpg', '9a85697fd693e11f87846ebc70c00227', 'Portrait of Marie Breunig', 'Realism', 'portrait', '1894'), ('100236.jpg', 'bf1d0c9118c5c4d6973bfeb5d7657d83', 'Omnibus', 'Impressionism', 'genre painting', '1893'), ('101455.jpg', 'baa3e6fec49e7eba24e23a8685c22955', 'Cosy Corner', 'Art Nouveau (Modern)', 'interior', '1894'), ('100845.jpg', 'acd9b34941d5485b6ffcf6836e07149f', 'The pilgrims resting', 'Biedermeier', 'genre painting', '1859'), ('100283.jpg', '3cc9a44380296d93e68b71a27643c25f', 'Road', 'Impressionism', 'landscape', '1876'), ('100152.jpg', '3cc9a44380296d93e68b71a27643c25f', 'Self-portrait in front of pink background', 'Impressionism', 'self-portrait', '1875'), ('100541.jpg', '6e606cd9cf9660466b2

In [ ]:
import csv

# Example dictionary
# filtered_data_by_filename

# Path where the CSV will be saved
csv_file_path = 'date_classification.csv'

# Create and write to the CSV file
with open(csv_file_path, mode='w', newline='') as file:
    writer = csv.writer(file)
    # Write the header
    writer.writerow(['filename', 'date range'])
    # Write data rows
    for date_range in new_data_by_date_range.keys():
      for image_info in new_data_by_date_range[date_range]:
          writer.writerow([image_info[0], date_range])

print(f"CSV file has been created at {csv_file_path}.")


CSV file has been created at date_classification.csv.


Create csv for style classification

In [ ]:
import csv

# Example dictionary
# filtered_data_by_filename

# Path where the CSV will be saved
csv_file_path = 'style_classification.csv'

# Create and write to the CSV file
with open(csv_file_path, mode='w', newline='') as file:
    writer = csv.writer(file)
    # Write the header
    writer.writerow(['filename', 'style'])
    # Write data rows
    for style in data_by_style.keys():
      for image_info in data_by_style[style]:
          writer.writerow([image_info[0], style])

print(f"CSV file has been created at {csv_file_path}.")


CSV file has been created at style_classification.csv.


In [ ]:
!ls

'CS 348K Final Project Notebook.ipynb'	'Kaggle JSON Folder'   train_2.zip      train_info.csv.zip
 date_classification.h5			 train_1	       train_3	       'Training Images'
 genre_classification.csv		 train_1.zip	       train_3.zip
 genre_classification.h5		 train_2	       train_info.csv


BUILDING Model Using Simple CNN

In [ ]:
# import tools
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
import matplotlib.pyplot as plt



Creating custom Data Generator

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

# Load your data
data = pd.read_csv('genre_classification.csv')

# Optionally shuffle the DataFrame if it's not already randomized
data = data.sample(frac=1).reset_index(drop=True)

# Split data into training and testing sets
train_data, test_data = train_test_split(data, test_size=0.15, random_state=42)  # 15% of the data is used for testing

# Separate filenames and labels for training and testing
train_filenames = train_data['filename'].values
train_labels = train_data['genre'].values

test_filenames = test_data['filename'].values
test_labels = test_data['genre'].values

# Encode labels
label_encoder = LabelEncoder()
train_labels_encoded = label_encoder.fit_transform(train_labels)
test_labels_encoded = label_encoder.transform(test_labels)  # Use the same encoder to ensure consistency


In [ ]:
import pandas as pd
import numpy as np
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.preprocessing.image import load_img, img_to_array
from sklearn.preprocessing import LabelEncoder


def image_generator(filenames, labels, batch_size, img_directory, target_size=(64, 64)):
    while True:  # Loop indefinitely
        for start in range(0, len(filenames), batch_size):
            end = min(start + batch_size, len(filenames))
            batch_filenames = filenames[start:end]
            batch_labels = labels[start:end]

            # Load images and process them
            images = [img_to_array(load_img(f"{img_directory}/{img_name}", target_size=target_size)) for img_name in batch_filenames]
            images = np.array(images) / 255.0  # Normalize the images

            # Convert labels to one-hot encoded
            batch_labels = to_categorical(batch_labels, num_classes=len(label_encoder.classes_))

            yield images, batch_labels

# Create generators
batch_size = 256
img_directory = 'Training Images - 256'
train_generator = image_generator(train_filenames, train_labels_encoded, batch_size, img_directory)
test_generator = image_generator(test_filenames, test_labels_encoded, batch_size, img_directory)


Configuring Model (CNN in Keras)

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

# Build a simple CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(64, 64, 3)),
    MaxPooling2D(2, 2),
    Conv2D(64, (3, 3), activation='relu'),
    # Removed one MaxPooling2D to preserve feature map size
    Flatten(),
    Dense(100, activation='relu'),
    Dropout(0.3),  # Adjusted dropout rate
    Dense(len(label_encoder.classes_), activation='softmax')
])


In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_2 (Conv2D)           (None, 62, 62, 32)        896       
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 31, 31, 32)        0         
 g2D)                                                            
                                                                 
 conv2d_3 (Conv2D)           (None, 29, 29, 64)        18496     
                                                                 
 flatten_1 (Flatten)         (None, 53824)             0         
                                                                 
 dense_2 (Dense)             (None, 100)               5382500   
                                                                 
 dropout_1 (Dropout)         (None, 100)               0         
                                                      

Train

In [ ]:
# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


In [ ]:
# This is the training for genre classification with train set 1, 2, and 3 from Kaggle

In [ ]:
model.fit(train_generator, steps_per_epoch=len(train_filenames) // batch_size, epochs=30)

Epoch 1/30
34/34 [==============================] - 33s 764ms/step - loss: 1.7936 - accuracy: 0.3653
Epoch 2/30
34/34 [==============================] - 27s 803ms/step - loss: 1.1036 - accuracy: 0.5224
Epoch 3/30
34/34 [==============================] - 28s 816ms/step - loss: 0.9801 - accuracy: 0.5859
Epoch 4/30
34/34 [==============================] - 26s 776ms/step - loss: 0.9013 - accuracy: 0.6297
Epoch 5/30
34/34 [==============================] - 26s 777ms/step - loss: 0.8632 - accuracy: 0.6450
Epoch 6/30
34/34 [==============================] - 25s 748ms/step - loss: 0.8040 - accuracy: 0.6751
Epoch 7/30
34/34 [==============================] - 27s 783ms/step - loss: 0.7568 - accuracy: 0.6939
Epoch 8/30
34/34 [==============================] - 27s 794ms/step - loss: 0.6987 - accuracy: 0.7249
Epoch 9/30
34/34 [==============================] - 27s 788ms/step - loss: 0.6534 - accuracy: 0.7491
Epoch 10/30
34/34 [==============================] - 27s 797ms/step - loss: 0.6207 - accura

KeyboardInterrupt: 

In [ ]:
model.fit(train_generator, steps_per_epoch=len(train_filenames) // batch_size, epochs=30)

Epoch 1/30
59/59 [==============================] - 10s 147ms/step - loss: 1.4267 - accuracy: 0.5587
Epoch 2/30
59/59 [==============================] - 8s 140ms/step - loss: 1.2394 - accuracy: 0.6111
Epoch 3/30
59/59 [==============================] - 11s 182ms/step - loss: 1.1326 - accuracy: 0.6417
Epoch 4/30
59/59 [==============================] - 9s 160ms/step - loss: 1.0806 - accuracy: 0.6610
Epoch 5/30
59/59 [==============================] - 8s 142ms/step - loss: 0.9951 - accuracy: 0.6754
Epoch 6/30
59/59 [==============================] - 9s 159ms/step - loss: 0.8639 - accuracy: 0.7102
Epoch 7/30
59/59 [==============================] - 9s 146ms/step - loss: 0.8409 - accuracy: 0.7311
Epoch 8/30
59/59 [==============================] - 9s 148ms/step - loss: 0.8028 - accuracy: 0.7300
Epoch 9/30
59/59 [==============================] - 9s 157ms/step - loss: 0.7627 - accuracy: 0.7440
Epoch 10/30
59/59 [==============================] - 8s 139ms/step - loss: 0.7013 - accuracy: 0.76

In [ ]:
print(len(os.listdir("Training Images")))

11109


In [ ]:
model.save('genre_classification_4_categories.h5')  # Saves the model in HDF5 format


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [ ]:
!ls

'CS 348K Final Project Notebook.ipynb'	'Kaggle JSON Folder'   train_info.csv.zip
 genre_classification.csv		 train_1	      'Training Images'
 genre_classification.gsheet		 train_1.zip
 genre_classification.h5		 train_info.csv


Testing the model for genre classification

In [ ]:
genre_model = models.load_model("genre_classification_4_categories.h5")

In [ ]:
genre_model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 62, 62, 32)        896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 31, 31, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 29, 29, 64)        18496     
                                                                 
 flatten (Flatten)           (None, 53824)             0         
                                                                 
 dense (Dense)               (None, 100)               5382500   
                                                                 
 dropout (Dropout)           (None, 100)               0         
                                                        

In [ ]:
test_loss, test_accuracy = genre_model.evaluate(test_generator, steps=len(test_filenames) // batch_size)

6/6 [==============================] - 327s 65s/step - loss: 1.3897 - accuracy: 0.2135


Let's test data by date range now:

Building Model for Date_Based Classification

In [ ]:
print(len(os.listdir("Training Images New")))

2484


In [ ]:
import csv
import os
import pandas as pd

In [ ]:
import csv

# Define the CSV file path
csv_file_path = 'date_classification.csv'

# Open the CSV file and count the number of rows
with open(csv_file_path, 'r', newline='') as csvfile:
    reader = csv.reader(csvfile)
    row_count = sum(1 for row in reader)

print(f"Number of rows in the CSV file: {row_count}")


Number of rows in the CSV file: 2485


In [ ]:
# import tools
import tensorflow as tf
from tensorflow.keras import datasets, layers, models
import matplotlib.pyplot as plt


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

# Load your data
data = pd.read_csv('date_classification.csv')

# Optionally shuffle the DataFrame if it's not already randomized
data = data.sample(frac=1).reset_index(drop=True)

# Split data into training and testing sets
train_data, test_data = train_test_split(data, test_size=0.15, random_state=42)  # 15% of the data is used for testing

# Separate filenames and labels for training and testing
train_filenames = train_data['filename'].values
train_labels = train_data['date range'].values

test_filenames = test_data['filename'].values
test_labels = test_data['date range'].values

# Encode labels
label_encoder = LabelEncoder()
train_labels_encoded = label_encoder.fit_transform(train_labels)
test_labels_encoded = label_encoder.transform(test_labels)  # Use the same encoder to ensure consistency


In [ ]:
import pandas as pd
import numpy as np
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.preprocessing.image import load_img, img_to_array
from sklearn.preprocessing import LabelEncoder

def image_generator(filenames, labels, batch_size, img_directory, target_size=(64, 64)):
    while True:  # Loop indefinitely
        for start in range(0, len(filenames), batch_size):
            end = min(start + batch_size, len(filenames))
            batch_filenames = filenames[start:end]
            batch_labels = labels[start:end]

            # Load images and process them
            images = [img_to_array(load_img(f"{img_directory}/{img_name}", target_size=target_size)) for img_name in batch_filenames]
            images = np.array(images) / 255.0  # Normalize the images

            # Convert labels to one-hot encoded
            batch_labels = to_categorical(batch_labels, num_classes=len(label_encoder.classes_))

            yield images, batch_labels

# Create generators
batch_size = 64
img_directory = 'Training Images New'
train_generator = image_generator(train_filenames, train_labels_encoded, batch_size, img_directory)
test_generator = image_generator(test_filenames, test_labels_encoded, batch_size, img_directory)


In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

# Build a simple CNN model
model = Sequential([
    Conv2D(32, (3, 3), activation='relu', input_shape=(64, 64, 3)),
    MaxPooling2D(2, 2),
    Conv2D(64, (3, 3), activation='relu'),
    # Removed one MaxPooling2D to preserve feature map size
    Flatten(),
    Dense(100, activation='relu'),
    Dropout(0.3),  # Adjusted dropout rate
    Dense(len(label_encoder.classes_), activation='softmax')
])


In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 62, 62, 32)        896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 31, 31, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 29, 29, 64)        18496     
                                                                 
 flatten (Flatten)           (None, 53824)             0         
                                                                 
 dense (Dense)               (None, 100)               5382500   
                                                                 
 dropout (Dropout)           (None, 100)               0         
                                                        

In [ ]:
# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])


In [ ]:
model.fit(train_generator, steps_per_epoch=len(train_filenames) // batch_size, epochs=30)

FileNotFoundError: [Errno 2] No such file or directory: 'Training Images New/25801.jpg'

In [ ]:
model.save('date_classification.h5')  # Saves the model in HDF5 format


/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [ ]:
test_loss, test_accuracy = model.evaluate(test_generator, steps=len(test_filenames) // batch_size)

4/4 [==============================] - 1s 144ms/step - loss: 5.7867 - accuracy: 0.3516


Training and testing on Style

Data loaded successfully.
Required columns are present.
Class distribution in the dataset:
style
Impressionism            246
Realism                  235
Romanticism              185
Expressionism            162
Post-Impressionism       140
                        ... 
Tubism                     1
Hyper-Realism              1
Byzantine                  1
Indian Space painting      1
Romanesque                 1
Name: count, Length: 105, dtype: int64
Error during train_test_split: Input contains NaN


NameError: name 'train_data' is not defined

In [ ]:
!ls

drive  sample_data
